File: python/tvm/relay/backend/contrib/uma/api/codegen.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Codegen base class of the Universal Modular Accelerator Interface (UMA)"""

from typing import Callable, Optional
import tvm


class UMACodegen(object):
    """
    Codegen base class of the Universal Modular Accelerator Interface (UMA)
    """

    def __init__(self, target_name: str) -> None:
        self.target_name = target_name

    def _register_codegen(
        self, fmt: str = "c", includes: Optional[Callable[[], str]] = None, **kwargs
    ) -> None:
        """Registration codegen in UMA.

        Parameters
        ----------
        fmt: str
            format of codegen. Currently only "c" is supported.
        includes : OptionalCallable[[], str]]
            user-defined function that adds C-#include statement to UMA C-Code.
        """
        if fmt == "c":
            self._register_c_codegen(includes, **kwargs)
        else:
            raise RuntimeError(f'Unsupported codegen format "{fmt}"')

    def _register_c_codegen(self, includes: Optional[Callable[[], str]] = None) -> None:
        """Registration of UMA helper functions, e.g. includes and replace_call_extern.

        Parameters
        ----------
        includes : OptionalCallable[[], str]]
            user-defined function that adds C-#include statement to UMA C-Code.
        """
        if includes is not None:
            tvm._ffi.register_func(
                f"relay.ext.uma.codegen_c_includes_{self.target_name}",
                includes,
                override=True,
            )

    def register(self) -> None:
        pass

File: src/relax/backend/contrib/xnnpack/codegen.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file src/relax/backend/contrib/xnnpack/codegen.cc
 * \brief Minimal XNNPACK Relax external codegen.
 */

#include <tvm/ffi/cast.h>
#include <tvm/ffi/extra/module.h>
#include <tvm/ffi/reflection/registry.h>
#include <tvm/ir/module.h>
#include <tvm/relax/expr.h>

#include <string>

#include "../codegen_json/codegen_json.h"
#include "../utils.h"

namespace tvm {
namespace relax {
namespace contrib {

using JSONGraphNode = tvm::runtime::json::JSONGraphNode;
using JSONSerializer = backend::contrib::JSONSerializer;
using backend::contrib::NodeEntries;

class XNNPACKJSONSerializer : public JSONSerializer {
 public:
  XNNPACKJSONSerializer(ffi::Map<Constant, ffi::String> constant_names,
                        ffi::Map<Var, Expr> bindings)
      : JSONSerializer(constant_names), bindings_(bindings) {}

  using JSONSerializer::VisitExpr_;

  NodeEntries VisitExpr_(const CallNode* call_node) final {
    const auto* fn_var = call_node->op.as<VarNode>();
    TVM_FFI_ICHECK(fn_var) << "XNNPACK codegen expects calls to composite functions.";

    const auto fn = Downcast<Function>(bindings_[ffi::GetRef<Var>(fn_var)]);
    TVM_FFI_ICHECK(fn.defined()) << "Expects the callee to be a function.";

    auto composite_opt = fn->GetAttr<ffi::String>(attr::kComposite);
    TVM_FFI_ICHECK(composite_opt.has_value()) << "Only composite functions are supported.";

    std::string composite_name = composite_opt.value();
    TVM_FFI_ICHECK_EQ(composite_name, "xnnpack.relu")
        << "Unsupported XNNPACK composite pattern: " << composite_name;

    NodeEntries inputs;
    for (const auto& arg : call_node->args) {
      auto res = VisitExpr(arg);
      inputs.insert(inputs.end(), res.begin(), res.end());
    }
    TVM_FFI_ICHECK_EQ(inputs.size(), 1U) << "xnnpack.relu expects exactly one input.";

    auto node = std::make_shared<JSONGraphNode>(composite_name, "kernel", inputs, 1);
    return AddNode(node, ffi::GetRef<Expr>(call_node));
  }

 private:
  ffi::Map<Var, Expr> bindings_;
};

ffi::Array<ffi::Module> XNNPACKCompiler(ffi::Array<Function> functions,
                                        ffi::Map<ffi::String, ffi::Any> /*options*/,
                                        ffi::Map<Constant, ffi::String> constant_names) {
  ffi::Array<ffi::Module> compiled_functions;
  const auto pf = tvm::ffi::Function::GetGlobalRequired("runtime.XNNPACKJSONRuntimeCreate");

  for (const auto& func : functions) {
    XNNPACKJSONSerializer serializer(constant_names, AnalyzeVar2Value(func));
    serializer.serialize(func);
    auto graph_json = serializer.GetJSON();
    auto const_names = serializer.GetConstantNames();
    auto func_name = GetExtSymbol(func);
    compiled_functions.push_back(pf(func_name, graph_json, const_names).cast<ffi::Module>());
  }

  return compiled_functions;
}

TVM_FFI_STATIC_INIT_BLOCK() {
  namespace refl = tvm::ffi::reflection;
  refl::GlobalDef().def("relax.ext.xnnpack", XNNPACKCompiler);
}

}  // namespace contrib
}  // namespace relax
}  // namespace tvm
